Failing build/release with numpy `1.24`:  `ValueError: setting an array element with a sequence`.

2023-07-06, two test which were previously passing are now failing. This is breaking the build.

Proposed fix in #3064 .

## The issue

These two tests are now failing:

```
test_tabular_multi_output_partition_masker:
test_tabular_multi_output_partition_masker:

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 2 dimensions. The detected shape was (12, 2) + inhomogeneous part.
```

Full traceback:

<details>

```
_________________ test_tabular_single_output_partition_masker __________________

    def test_tabular_single_output_partition_masker():
        model, data = common.basic_xgboost_scenario(100)
>       common.test_additivity(shap.explainers.Exact, model.predict, shap.maskers.Partition(data), data)

tests/explainers/test_exact.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/explainers/common.py:31: in test_additivity
    explainer = explainer_type(model, masker, **kwargs)
/opt/hostedtoolcache/Python/3.11.4/x[64](https://github.com/slundberg/shap/actions/runs/5466769146/jobs/9955997146#step:6:65)/lib/python3.11/site-packages/shap/explainers/_exact.py:[65](https://github.com/slundberg/shap/actions/runs/5466769146/jobs/9955997146#step:6:66): in __init__
    self._partition_masks, self._partition_masks_inds = partition_masks(masker.clustering)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

partition_tree = array([[ 6.        , 11.        ,  0.70911646,  2.        ],
       [ 2.        , 10.        ,  0.73290[67](https://github.com/slundberg/shap/actions/runs/5466769146/jobs/9955997146#step:6:68)2,  2.       ...    [ 3.        , 20.        ,  0.892[73](https://github.com/slundberg/shap/actions/runs/5466769146/jobs/9955997146#step:6:74)649, 11.        ],
       [ 8.        , 21.        ,  0.903310[78](https://github.com/slundberg/shap/actions/runs/5466769146/jobs/9955997146#step:6:79), 12.        ]])

    def partition_masks(partition_tree):
        """ Return an array of all the masks possible while following the given partition tree.
        """
    
        M = partition_tree.shape[0] + 1
        mask_matrix = make_masks(partition_tree)
        all_masks = []
        m00 = np.zeros(M, dtype=bool)
        all_masks.append(m00)
        all_masks.append(~m00)
        #inds_stack = [0,1]
        inds_lists = [[[], []] for i in range(M)]
        _partition_masks_recurse(len(partition_tree)-1, m00, 0, 1, inds_lists, mask_matrix, partition_tree, M, all_masks)
    
        all_masks = np.array(all_masks)
    
        # we resort the clustering matrix to minimize the sequential difference between the masks
        # this minimizes the number of model evaluations we need to run when the background sometimes
        # matches the foreground. We seem to average about 1.5 feature changes per mask with this
        # approach. This is not as clean as the grey code ordering, but a perfect 1 feature change
        # ordering is not possible with a clustering tree
        order = delta_minimization_order(all_masks)
        inverse_order = np.arange(len(order))[np.argsort(order)]
    
        for inds_list0,inds_list1 in inds_lists:
            for i in range(len(inds_list0)):
                inds_list0[i] = inverse_order[inds_list0[i]]
                inds_list1[i] = inverse_order[inds_list1[i]]
    
>       return all_masks[order], np.array([[np.array(on), np.array(off)] for on,off in inds_lists])
E       ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 2 dimensions. The detected shape was (12, 2) + inhomogeneous part.
```

</details>

## Cause

Tensorflow 2.13.0 was just released. This in turn bumped us up to numpy 1.24, which contains a breaking change.

Here's a comparison of the versions installed by pip on CI, before and after the tests began to fail:

<details>

```
# Before
Successfully installed Jinja2-3.1.2 MarkupSafe-2.1.3 absl-py-1.4.0 asttokens-2.2.1 astunparse-1.6.3 backcall-0.2.0 cachetools-5.3.1 catboost-1.2 certifi-2023.5.7 charset-normalizer-3.1.0 cloudpickle-2.2.1 cmake-3.26.4 contourpy-1.1.0 coverage-7.2.7 cycler-0.11.0 decorator-5.1.1 executing-1.2.0 filelock-3.12.2 flatbuffers-23.5.26 fonttools-4.40.0 fsspec-2023.6.0 gast-0.4.0 google-auth-2.21.0 google-auth-oauthlib-1.0.0 google-pasta-0.2.0 graphviz-0.20.1 grpcio-1.56.0 h5py-3.9.0 huggingface-hub-0.15.1 idna-3.4 iniconfig-2.0.0 ipython-8.14.0 jax-0.4.13 jedi-0.18.2 joblib-1.3.1 keras-2.12.0 kiwisolver-1.4.4 libclang-16.0.0 lightgbm-3.3.5 lit-16.0.6 llvmlite-0.40.1 markdown-3.4.3 matplotlib-3.7.1 matplotlib-inline-0.1.6 ml-dtypes-0.2.0 mpmath-1.3.0 networkx-3.1 numba-0.57.1 numpy-1.23.5 nvidia-cublas-cu11-11.10.3.66 nvidia-cuda-cupti-cu11-11.7.101 nvidia-cuda-nvrtc-cu11-11.7.99 nvidia-cuda-runtime-cu11-11.7.99 nvidia-cudnn-cu11-8.5.0.96 nvidia-cufft-cu11-10.9.0.58 nvidia-curand-cu11-10.2.10.91 nvidia-cusolver-cu11-11.4.0.1 nvidia-cusparse-cu11-11.7.4.91 nvidia-nccl-cu11-2.14.3 nvidia-nvtx-cu11-11.7.91 oauthlib-3.2.2 opencv-python-4.8.0.74 opt-einsum-3.3.0 packaging-23.1 pandas-2.0.3 parso-0.8.3 pexpect-4.8.0 pickleshare-0.7.5 pillow-10.0.0 plotly-5.15.0 pluggy-1.2.0 prompt-toolkit-3.0.39 protobuf-3.20.3 ptyprocess-0.7.0 pure-eval-0.2.2 py4j-0.10.9.7 pyasn1-0.5.0 pyasn1-modules-0.3.0 pygments-2.15.1 pyod-1.1.0 pyparsing-3.1.0 pyspark-3.4.1 pytest-7.4.0 pytest-cov-4.1.0 pytest-mpl-0.16.1 python-dateutil-2.8.2 pytz-2023.3 pyyaml-6.0 regex-2023.6.3 requests-2.31.0 requests-oauthlib-1.3.1 rsa-4.9 safetensors-0.3.1 scikit-learn-1.3.0 scipy-1.11.1 sentencepiece-0.1.99 setuptools-68.0.0 shap-0.42.0 six-1.16.0 slicer-0.0.7 stack-data-0.6.2 sympy-1.12 tenacity-8.2.2 tensorboard-2.12.3 tensorboard-data-server-0.7.1 tensorflow-2.12.0 tensorflow-estimator-2.12.0 tensorflow-io-gcs-filesystem-0.32.0 termcolor-2.3.0 threadpoolctl-3.1.0 tokenizers-0.13.3 torch-2.0.1 torchvision-0.15.2 tqdm-4.65.0 traitlets-5.9.0 transformers-4.30.2 triton-2.0.0 typing-extensions-4.7.1 tzdata-2023.3 urllib3-1.26.16 wcwidth-0.2.6 werkzeug-2.3.6 wheel-0.40.0 wrapt-1.14.1 xgboost-1.7.6

# After
Successfully installed Jinja2-3.1.2 MarkupSafe-2.1.3 absl-py-1.4.0 asttokens-2.2.1 astunparse-1.6.3 backcall-0.2.0 cachetools-5.3.1 catboost-1.2 certifi-2023.5.7 charset-normalizer-3.1.0 cloudpickle-2.2.1 cmake-3.26.4 contourpy-1.1.0 coverage-7.2.7 cycler-0.11.0 decorator-5.1.1 executing-1.2.0 filelock-3.12.2 flatbuffers-23.5.26 fonttools-4.40.0 fsspec-2023.6.0 gast-0.4.0 google-auth-2.21.0 google-auth-oauthlib-1.0.0 google-pasta-0.2.0 graphviz-0.20.1 grpcio-1.56.0 h5py-3.9.0 huggingface-hub-0.16.2 idna-3.4 iniconfig-2.0.0 ipython-8.14.0 jedi-0.18.2 joblib-1.3.1 keras-2.13.1 kiwisolver-1.4.4 libclang-16.0.0 lightgbm-3.3.5 lit-16.0.6 llvmlite-0.40.1 markdown-3.4.3 matplotlib-3.7.1 matplotlib-inline-0.1.6 mpmath-1.3.0 networkx-3.1 numba-0.57.1 numpy-1.24.3 oauthlib-3.2.2 opencv-python-4.8.0.74 opt-einsum-3.3.0 packaging-23.1 pandas-2.0.3 parso-0.8.3 pexpect-4.8.0 pickleshare-0.7.5 pillow-10.0.0 plotly-5.15.0 pluggy-1.2.0 prompt-toolkit-3.0.39 protobuf-3.20.3 ptyprocess-0.7.0 pure-eval-0.2.2 py4j-0.10.9.7 pyasn1-0.5.0 pyasn1-modules-0.3.0 pygments-2.15.1 pyod-1.1.0 pyparsing-3.1.0 pytest-7.4.0 pytest-cov-4.1.0 pytest-mpl-0.16.1 python-dateutil-2.8.2 pytz-2023.3 pyyaml-6.0 regex-2023.6.3 requests-2.31.0 requests-oauthlib-1.3.1 rsa-4.9 safetensors-0.3.1 scikit-learn-1.3.0 scipy-1.11.1 sentencepiece-0.1.99 setuptools-68.0.0 shap-0.42.0 six-1.16.0 slicer-0.0.7 stack-data-0.6.2 sympy-1.12 tenacity-8.2.2 tensorboard-2.13.0 tensorboard-data-server-0.7.1 tensorflow-2.13.0 tensorflow-estimator-2.13.0 tensorflow-io-gcs-filesystem-0.32.0 termcolor-2.3.0 threadpoolctl-3.1.0 tokenizers-0.13.3 tqdm-4.65.0 traitlets-5.9.0 transformers-4.30.2 triton-2.0.0 typing-extensions-4.5.0 tzdata-2023.3 urllib3-1.26.16 wcwidth-0.2.6 werkzeug-2.3.6 wheel-0.40.0 wrapt-1.15.0 xgboost-1.7.6

# Diff

< numpy-1.23.5
> numpy-1.24.3

< huggingface-hub-0.15.1
> huggingface-hub-0.16.2

< keras-2.12.0
> keras-2.13.1

< tensorflow-2.12.0
< tensorflow-estimator-2.12.0
< tensorboard-2.12.3
> tensorflow-2.13.0
> tensorflow-estimator-2.13.0
> tensorboard-2.13.0

< typing-extensions-4.7.1
> typing-extensions-4.5.0

< wrapt-1.14.1
> wrapt-1.15.0

< pyspark-3.4.1
< torch-2.0.1
< torchvision-0.15.2
< ml-dtypes-0.2.0
< jax-0.4.13

```

</details>

I believe the relevant change in the numpy [1.24 release notes](https://numpy.org/devdocs/release/1.24.0-notes.html) is:

> Ragged array creation will now always raise a ValueError unless dtype=object is passed

The array `Exact._partition_masks_inds` is a ragged array, as the sub-elements have different lengths. It's created here:

https://github.com/slundberg/shap/blob/dd63adb842cf064b0085120dd4e664850453d720/shap/explainers/_exact.py#L289

I notice there was a DeprecationWarning in our previously passing test suite to this effect, which hadn't yet been fixed. This issue is perhaps a good call-to-arms for us to address the remaining deprecation warnings, before they lead to similar failures in future.

```
tests/explainers/test_exact.py::test_tabular_single_output_partition_masker
tests/explainers/test_exact.py::test_tabular_multi_output_partition_masker
  Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -66,6 +66,8 @@ into the main shap repository. PRs from this origin are labelled here as `fork#1
   [#3028](https://github.com/slundberg/shap/pull/3028),
   [#3029](https://github.com/slundberg/shap/pull/3029) and
   [#3031](https://github.com/slundberg/shap/pull/3031) by @connortann).
+- Fixed creation of ragged arrays in `shap.explainers.Exact`
+  ([#3064](https://github.com/slundberg/shap/pull/3064) by @connortann).
 
 ### Changed
 

--- a/shap/explainers/_exact.py
+++ b/shap/explainers/_exact.py
@@ -286,7 +286,10 @@ def partition_masks(partition_tree):
             inds_list0[i] = inverse_order[inds_list0[i]]
             inds_list1[i] = inverse_order[inds_list1[i]]
 
-    return all_masks[order], np.array([[np.array(on), np.array(off)] for on,off in inds_lists])
+    # Care: inds_lists have different lengths, so partition_masks_inds is a "ragged" array. See GH #3063
+    partition_masks = all_masks[order]
+    partition_masks_inds = [[np.array(on), np.array(off)] for on, off in inds_lists]
+    return partition_masks, partition_masks_inds
 
 # TODO: this should be a jit function... which would require preallocating the inds_lists (sizes are 2**depth of that ind)
 # TODO: we could also probable avoid making the masks at all and just record the deltas if we want...

